File: src/scitrader.marketdataserver/src/main/java/com/scitrader/marketdataserver/exchange/bitmex/Tick.java
package com.scitrader.marketdataserver.exchange.bitmex;

public class Tick {
  private String timestamp;
  private String symbol;
  private String side;
  private float size;
  private float price;
  private String tickDirection;
  private String trdMatchID;
  private float grossValue;
  private float homeNotional;
  private float foreignNotional;


  // Getter Methods

  public String getTimestamp() {
    return timestamp;
  }

  public String getSymbol() {
    return symbol;
  }

  public String getSide() {
    return side;
  }

  public float getSize() {
    return size;
  }

  public float getPrice() {
    return price;
  }

  public String getTickDirection() {
    return tickDirection;
  }

  public String getTrdMatchID() {
    return trdMatchID;
  }

  public float getGrossValue() {
    return grossValue;
  }

  public float getHomeNotional() {
    return homeNotional;
  }

  public float getForeignNotional() {
    return foreignNotional;
  }

  // Setter Methods

  public void setTimestamp(String timestamp) {
    this.timestamp = timestamp;
  }

  public void setSymbol(String symbol) {
    this.symbol = symbol;
  }

  public void setSide(String side) {
    this.side = side;
  }

  public void setSize(float size) {
    this.size = size;
  }

  public void setPrice(float price) {
    this.price = price;
  }

  public void setTickDirection(String tickDirection) {
    this.tickDirection = tickDirection;
  }

  public void setTrdMatchID(String trdMatchID) {
    this.trdMatchID = trdMatchID;
  }

  public void setGrossValue(float grossValue) {
    this.grossValue = grossValue;
  }

  public void setHomeNotional(float homeNotional) {
    this.homeNotional = homeNotional;
  }

  public void setForeignNotional(float foreignNotional) {
    this.foreignNotional = foreignNotional;
  }

  @Override
  public String toString() {
    return String.format("%s, %s, %s at %s on %s", getSymbol(), getSide(), getSize(), getPrice(), getTimestamp());
  }
}

File: Kotlin/hello-kotlin-world/src/main/java/com/scitrader/marketdataserver/exchange/bitmex/messages/Tick.kt
package com.scitrader.marketdataserver.exchange.bitmex.messages

import com.jsoniter.annotation.JsonIgnore
import com.scitrader.marketdataserver.common.Utility.DateUtil
import org.bson.Document
import org.joda.time.DateTime
import java.util.*

class Tick {
    @JsonIgnore
    internal var dateTime: DateTime? = null

    var timestamp: String? = null
    var symbol: String? = null
    var side: String? = null
    var size: Float = 0.toFloat()
    var price: Float = 0.toFloat()
    var tickDirection: String? = null
    var trdMatchID: String? = null
    var grossValue: Float = 0.toFloat()
    var homeNotional: Float = 0.toFloat()
    var foreignNotional: Float = 0.toFloat()

    // e.g. 2019-01-04T12:17:00.980Z
    var timeStampAsDate: DateTime?
        get() {

            if (dateTime == null && timestamp != null) {
                dateTime = DateUtil.tickFormatter.parseDateTime(timestamp)
            }
            return dateTime
        }
        set(dateTime) {
            this.dateTime = dateTime
        }

    override fun toString(): String {
        return String.format("%s, %s, %s at %s on %s", symbol, side, size, price, timestamp)
    }

    fun toBsonDocument(): Document {
        val doc = Document()

        doc.append("_id", trdMatchID)
        doc.append("Time", timeStampAsDate!!.millis)
        doc.append("Price", price)
        doc.append("Size", size * if (side!!.toUpperCase(Locale.US) === "BUY") 1f else -1f)

        return doc
    }

    companion object {

        fun fromBsonDocument(symbol: String, document: Document): Tick {
            val t = Tick()
            t.symbol = symbol
            t.size = (document["Size"] as Double).toFloat()
            t.price = (document["Price"] as Double).toFloat()
            t.timeStampAsDate = DateTime(document["Time"] as Long)
            t.side = if (t.size < 0) "Sell" else "Buy"
            return t
        }
    }
}